plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
+	// gather CPU metrics relying on coreFreq and msr which share CPU IDs.
+	if p.needsCoreFreq || p.needsMsrCPU {
+		p.addCPUMetrics(acc)
+	}
+
+	// gather CPU metrics relying on perf.
+	if p.needsPerf {
+		p.addCPUPerfMetrics(acc)
+	}
 
-		p.cpuBusClockValue = busClockInfo
+	// gather package metrics.
+	if len(p.PackageMetrics) != 0 {
+		p.addPackageMetrics(acc)
 	}
+
+	return nil
 }
 
-// Gather takes in an accumulator and adds the metrics that the Input gathers
-func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
-	if p.areGlobalMetricsEnabled() {
-		p.addGlobalMetrics(acc)
+// parseConfig is a helper method that parses configuration fields from the receiver such as included/excluded CPU IDs.
+func (p *PowerStat) parseConfig() error {
+	var err error
+
+	if p.MsrReadTimeout < 0 {
+		return errors.New("msr_read_timeout should be positive number or equal to 0 (to disable timeouts)")
 	}
 
-	if p.areCoreMetricsEnabled() {
-		if p.msr.isMsrLoaded() {
-			p.logOnce["msr"] = nil
-			p.addPerCoreMetrics(acc)
-		} else {
-			err := errors.New("Error while trying to read MSR (probably msr module was not loaded)")
-			if val := p.logOnce["msr"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("%v", err)
-				// Remember that specific error occurs to omit logging next time
-				p.logOnce["msr"] = err
-			}
-		}
+	p.PackageMetrics, err = parsePackageMetrics(p.PackageMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to parse package metrics: %w", err)
 	}
 
-	// Gathering the first iteration of metrics was skipped for most of them because they are based on delta calculations
-	p.skipFirstIteration = false
+	p.CPUMetrics, err = parseCPUMetrics(p.CPUMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to parse core metrics: %w", err)
+	}
 
-	return nil
-}
+	if len(p.CPUMetrics) == 0 && len(p.PackageMetrics) == 0 {
+		return fmt.Errorf("no metrics were found in the configuration file")
+	}
 
-func (p *PowerStat) addGlobalMetrics(acc telegraf.Accumulator) {
-	// Prepare RAPL data each gather because there is a possibility to disable rapl kernel module
-	p.rapl.initializeRaplData()
-	for socketID := range p.rapl.getRaplData() {
-		if p.packageTurboLimit {
-			p.addTurboRatioLimit(socketID, acc)
-		}
+	p.parsedCPUTimedMsrMetrics = parseCPUTimeRelatedMsrMetrics(p.CPUMetrics)
+	p.parsedCPUPerfMetrics = parseCPUPerfMetrics(p.CPUMetrics)
 
-		if p.packageUncoreFrequency {
-			die := maxDiePerSocket(socketID)
-			for actualDie := 0; actualDie < die; actualDie++ {
-				p.addUncoreFreq(socketID, strconv.Itoa(actualDie), acc)
-			}
-		}
+	p.parsedPackageRaplMetrics = parsePackageRaplMetrics(p.PackageMetrics)
+	p.parsedPackageMsrMetrics = parsePackageMsrMetrics(p.PackageMetrics)
 
-		if p.packageCPUBaseFrequency {
-			p.addCPUBaseFreq(socketID, acc)
-		}
+	if len(p.ExcludedCPUs) != 0 && len(p.IncludedCPUs) != 0 {
+		return errors.New("configuration error. Provide either one 'included_cpus'/'excluded_cpus' configuration option, or none")
```

Comment:
How about
```suggestion
		return errors.New("both 'included_cpus' and 'excluded_cpus' configured; provide only one or none of the two")
```

Alternatively, you could handle the case where both are specified by first gathering the included **and** the excluded CPUs, Then, in a second step, remove the excluded CPUs from the include-list and specify the result as `parsedIncludedCPUs`, leaving the `parsedExcludedCores` empty...

Reply:
We would like to handle the simpler scenario by returning error if both options are used in config file. I believe changing the error message as you suggested is the way to go.